File: 00199. Binary Tree Right Side View/00199.c
#define MAX_NODE_DEPTH  100
#define INIT_VAL        101

void traversal(struct TreeNode *root, int depth, char tmp[]) {
    if (root == NULL) {
        return;
    }
    
    tmp[depth] = root->val;
    traversal(root->left,  depth+1, tmp);
    traversal(root->right, depth+1, tmp);
}

int * rightSideView(struct TreeNode *root, int *returnSize) {
    static int ret[MAX_NODE_DEPTH] = {};
    char tmp[MAX_NODE_DEPTH] = {};
    memset(ret, 0, sizeof(int) * MAX_NODE_DEPTH);
    memset(tmp, INIT_VAL, sizeof(char) * MAX_NODE_DEPTH);
    
    traversal(root, 0, tmp);

    int i;
    
    for (i = 0; i < MAX_NODE_DEPTH && tmp[i] != INIT_VAL; ++i) {
        ret[i] = (int)(tmp[i]);
    }
    *returnSize = i;
    
    return ret;
}

File: LeetCode/00199. Binary Tree Right Side View/00199.java
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
  
        traversal(result, root, 0);
        
        return result;
    }
    
    public void traversal(List<Integer> result, TreeNode root, int depth) {
        if (root == null) {
            return;
        }
        
        if (result.size() == depth) {
            result.add(root.val);
        }
        
        traversal(result, root.right, depth+1);
        traversal(result, root.left,  depth+1);
    }
}
